Mathlib/RingTheory/Coalgebra.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2023 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Coalgebras
+
+In this file we define `Coalgebra`, and provide instances for:
+
+* Commutative rings: `CommRing.toCoalgebra`
+* Finitely supported functions: `Finsupp.instCoalgebra`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Coalgebra>
+-/
+
+suppress_compilation
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A coalgebra over a commutative ring `R` is a module over `R` equipped with a coassociative
+comultiplication `Δ` and a counit `ε` obeying the left and right conunitality laws. -/
+class Coalgebra (R : Type u) (A : Type v) [CommRing R] [AddCommGroup A] [Module R A] where
+  /-- The comultiplication of the coalgebra -/
+  comul : A →ₗ[R] A ⊗[R] A
+  /-- The counit of the coalgebra -/
+  counit : A →ₗ[R] R
+  /-- The comultiplication is coassociative -/
+  coassoc : TensorProduct.assoc R A A A ∘ₗ comul.rTensor A ∘ₗ comul = comul.lTensor A ∘ₗ comul
+  /-- The counit satisfies the left counitality law -/
+  counit_id : TensorProduct.lid R A ∘ₗ counit.rTensor A ∘ₗ comul = .id
+  /-- The counit satisfies the right counitality law -/
+  id_counit : TensorProduct.rid R A ∘ₗ counit.lTensor A ∘ₗ comul = .id
```

Comment:
Note that https://github.com/leanprover-community/mathlib4/pull/8621#discussion_r1408558962 still applies here, as
```suggestion
  /-- The counit satisfies the left counitality law -/
  counit_id : counit.rTensor A ∘ₗ comul = TensorProduct.mk R _ _ 1
  /-- The counit satisfies the right counitality law -/
  id_counit : counit.lTensor A ∘ₗ comul = (TensorProduct.mk R _ _).flip 1
```

At this point it would probably make sense to rename the fields to` rTensor_counit_comp_comul` and `lTensor_counit_comp_comul`

Reply:
When I change these, should I change the statements of the lemmas that apply them, or keep the same statements and change the proofs?

Reply:
You should probably do both: add new lemmas with the new statements (which probably work with effectively the same old proofs), but also keep the old ones, with new proofs. The old statements should follow trivially from the new ones